### ISSUE TITLE
TypeError with download path on Windows

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#2305]
Report on v0.8.4, on Windows:

```
09:18:06 DEBUG    downloads  downloads:fetch:879 fetch: PyQt5.QtCore.QUrl('http://www.uniklinikum-regensburg.de/imperia/md/content/kliniken-institute/psychosomatischemedizin/studenten/psycheundumweltbelastung.pdf') -> psycheundumweltbelastung.pdf
09:18:06 DEBUG    misc       standarddir:_writable_location:128 writable location for DownloadLocation: H:/Eigene Dateien/Downloads
09:18:06 DEBUG    statusbar  prompter:ask_question:340 Asking question <qutebrowser.utils.usertypes.Question default='H:\\Eigene Dateien\\Downloads\\' mode=<PromptMode.download: 5> text='Save file to:'>, blocking False, loops [], queue deque([])
# [...]
09:18:13 ERROR    message    message:error:307 XDG_DOWNLOAD_DIR points to a relative path - please check your ~/.config/user-dirs.dirs. The download is saved in your home directory.
# [...]
09:18:13 ERROR    misc       crashsignal:exception_hook:211 Uncaught exception
Traceback (most recent call last):
  File "scripts/dev\..\..\scripts\..\qutebrowser\browser\webkit\downloads.py", line 947, in _set_download_target
  File "scripts/dev\..\..\scripts\..\qutebrowser\browser\webkit\downloads.py", line 580, in set_filename
  File "c:\Python\64-bit\3.4\lib\ntpath.py", line 246, in basename
  File "c:\Python\64-bit\3.4\lib\ntpath.py", line 217, in split
  File "c:\Python\64-bit\3.4\lib\ntpath.py", line 159, in splitdrive
TypeError: object of type 'NoneType' has no len()
```

Not sure what triggered that error on Windows...

**Comment by Kingdread** [comment]
Okay, I've done some digging, and I've stumbled upon some quite interesting behaviour:

So first of all, the bug doesn't seem to exist on the latest version anymore. The message still appears, but due to a change of how the fallback is implemented, we actually get a valid download directory. Namely, we use `create_full_filename(self.basename, os.path.expanduser('~'))` instead of `create_full_filename(self.basename, os.path.expanduser(os.path.join('~', filename)))`, which in this case prevents the crash.

The bug occurs when giving `E:` (or some other drive) as download directory: The correct way to specify a drive would be `E:\` with the backslash (`>>> os.path.isabs("E:\\") -> True`), but when we omit the backslash, we don't actually mean the drive `E:` (`>>> os.path.isabs("E:") -> False`). Why not? Because NTFS supports *alternate data streams* in a file, and those streams are denoted by a colon. Thus `E:stream` would actually refer to the stream named "stream" in the *relative* file "E" (and I guess the Python devs took care of this).

Now that wouldn't be a problem, since we prepend the download directory in case of a relative file, but as it turns out, `os.path.join` handles things a bit differently: Even though `E:` is seen as a relative path, joining it treats it as an absolute path:

```
>>> os.path.join("C:\\Eigene Dateien\\Downloads", "E:")
'E:'
```

So - `E:` is not an absolute path, but we also can't turn it into an absolute path by prepending the download dir, as `E:` behaves like an absolute path here. I'm not sure if this could qualify as a bug in the Python stdlib (and I'm not enough of a Windows expert to tell), but it sure as hell is confusing. Our whole directory logic breaks down because we can't turn `E:` into an absolute path.

Since alternate streams are rarely (if ever) used as a download target (and it seems like the "empty" stream just refers to the file itself), I think we can change the behaviour so that the `X:` path is automatically turned into the (correct) `X:\` path, and I guess that that behaviour is expected by users (because nobody actually wants to type the backslash).

**Comment by The-Compiler** [comment]
The whole idea of having a working-directory per-drive is ~~dumb~~ exciting! 😆 Thanks for tracking this down! [This Python issue](http://bugs.python.org/issue23535) has some background too (at least it got fixed in Python 3.4, so we don't have to take care of different behavior across Python versions...).

So what do you think we should do here? Do we get some sensible directory now? Should we just take care of the message, write a test and be done with it?

**Comment by Kingdread** [comment]
Currently, we just fall back to the user's home directory, so if he types `E:`, it will be saved in ` C:\Users\...`.

I propose to special-case the behaviour so that `E:` is treated the same as `E:\` - after all, that's probably what most users expect, I doubt anyone really means the "working directory on `E:`" (especially since you cannot change it from within qute and we disallow relative paths anyway).
